modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectChangeFieldTypeTest.java

```diff
@@ -0,0 +1,266 @@
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionIsolation;
+import org.junit.Test;
+
+/**
+ *
+ */
+public class BinaryObjectChangeFieldTypeTest extends GridCommonAbstractTest {
+
+    /** */
+    private static final int NODE_COUNT = 2;
+
+    /** Cache name */
+    private static final String CACHE_NAME = "CACHE";
+
+    /** */
+    private IgniteTransactions transactions;
+    /** */
+    private IgniteEx ignite;
+    /** */
+    private IgniteBinary binary;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override protected void beforeTest() throws Exception {
+        startGrids(NODE_COUNT);
+
+        super.beforeTest();
+
+        ignite = grid(0);
+
+        binary = ignite.binary();
+
+        transactions = ignite.transactions();
+    }
+
+    /**
+     * {@inheritDoc}
+     * */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setMarshaller(new BinaryMarshaller());
+
+        return cfg;
+    }
+
+    /**
+     * Check commit changed BinaryObject to non-backup cache
+     */
+    @Test
+    public void testTxNoBackups() {
+
+        CacheConfiguration<Integer, TestClass> cc = new CacheConfiguration<Integer, TestClass>(CACHE_NAME)
+            .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+            .setCacheMode(CacheMode.PARTITIONED)
+            .setIndexedTypes(Integer.class, TestClass.class);
+
+        IgniteCache<Integer, TestClass> fooCache = ignite.getOrCreateCache(cc);
+
+        IgniteCache<Integer, BinaryObject> binCache = fooCache.withKeepBinary();
+
+        try (Transaction tx = transactions.txStart()) {
+            fooCache.put(1, new TestClass("dummy", 123));
+            tx.commit();
+        }
+
+        assertEquals(123, fooCache.get(1).intField);
+
+        try (Transaction tx = transactions.txStart()) {
+            fooCache.put(1, new TestClass("dummy", 5));
+
+            BinaryObject bo = binary.builder(TestClass.class.getName())
+                .build()
+                .toBuilder()
+                .setField("intField", "String")
+                .build();
+
+            binCache.put(1001, bo);
+
+            tx.commit();
+            fail("exception expected!!!");
+        }
+        catch (Exception e) {
+            assertTrue(e.getMessage().startsWith("Wrong value has been set"));
+        }
+
+        assertEquals(123, fooCache.get(1).intField);
```

Comment:
i think this check is excessive , if u remove it u can use something like : @Test(expected = NullPointerException.class)
just check for appropriate message ant throw this exception further, would be more clear i hope.